blocks/bricks/recurrent.py

```diff
@@ -630,3 +631,140 @@ def apply(self, *args, **kwargs):
     @apply.delegate
     def apply_delegate(self):
         return self.children[0].apply
+
+
+class RecurrentStack(BaseRecurrent, Initializable):
+    u"""Stack of recurrent networks.
+
+    Build a stack of recurrent layers from a supplied list of
+    BaseRecurrent objects. Each object must have a `sequences`,
+    `contexts`, `states` and `outputs` parameters to its `apply` method.
+    It is assumed that the `sequences` of all objects has an "inputs"
+    element. It is assumed that all `states` have a "states" element.
+
+    The `sequences` (the "inputs") of the the first layer (layer 0) is
+    used as the `sequence` of the stack (the instance of this class.)
+    The "mask" element of the `sequences` is carried to all layers that
+    have a "mask" element in their `sequences`.
+    For all the following layers, all elements of the `sequences`, except
+    for "mask", are forked from the "states" element of the of the
+    `states` of the previous layer.
+
+    The concatenated `states` of all layers is the `states` of the stack.
+    In order to avoid conflict, the names of the elements appearing in the
+    `states` of each layers are suffixed with a '_<n>' before combining
+    them. The '<n>' is the number of the layer starting from 0.
+    The same process is true for the `contexts` and `outputs` of the
+    stack.
+
+    Parameters
+    ----------
+    transitions : list
+        List of recurrent units to use in each layer.
+    prototype : :class:`.FeedForward`, optional
+        The transformation applied to each states of each layer  when it
+        is forked to the sequences of the next layer. By
+        default a :class:`.Linear` transformation with biases is used.
+
+    Notes
+    -----
+    See :class:`.BaseRecurrent` for more initialization parameters.
+
+    """
+    def __init__(self, transitions, prototype=None, **kwargs):
+        super(RecurrentStack, self).__init__(**kwargs)
+
+        for d, transition in enumerate(transitions):
+            transition.name += '_%d' % d
+        self.transitions = transitions
+
+        if prototype is None:
+            # By default use Linear (with bias) as a default fork.
+            # This overrides the bad default inside Fork which is without bias.
+            # Yes I know use_bias=True us the defaut, but it is a big deal so
+            # I write it down explicitly.
+            prototype = Linear(use_bias=True)
+        depth = len(transitions)
+        self.forks = [Fork(self.normal_inputs(d), prototype=prototype)
+                      for d in range(1, depth)]
+
+        self.children = self.transitions + self.forks
+
+        # Programmatically fix the parameters of the  recurrent decoration
+        setattr(self.apply, 'sequences', transitions[0].apply.sequences)
```

Comment:
Cool, this is a new technique which I never though of.. and I like it!

The standard way to have application method properties defined at runtime was so far

```
@apply.property('inputs')
def apply_inputs(self):
    return self.wrapped.inputs
```

but seems like simply setting attributes of the `BoundApplication` also does the job!